src/libraries/System.Formats.Tar/tests/TarReader/TarReader.File.Async.Tests.cs

```diff
@@ -125,5 +126,245 @@ public class TarReader_File_Async_Tests : TarReader_File_Tests_Async_Base
+
+        [Theory]
+        [MemberData(nameof(GetPaxAndGnuTestCaseNames))]
+        public Task ReadCopiedDataStreamOfTarGzOldGnuAsync(string testCaseName) =>
+            VerifyDataStreamOfTarGzInternalAsync(TestTarFormat.oldgnu, testCaseName, copyData: true);
+
+        [Theory]
+        [MemberData(nameof(GetPaxAndGnuTestCaseNames))]
+        public Task ReadCopiedDataStreamOfTarGzGnuAsync(string testCaseName) =>
+            VerifyDataStreamOfTarGzInternalAsync(TestTarFormat.gnu, testCaseName, copyData: true);
+
+        [Theory]
+        [MemberData(nameof(GetGoLangTarTestCaseNames))]
+        public Task ReadDataStreamOfExternalAssetsGoLangAsync(string testCaseName) =>
+            VerifyDataStreamOfTarUncompressedInternalAsync("golang_tar", testCaseName, copyData: false);
+
+        [Theory]
+        [MemberData(nameof(GetNodeTarTestCaseNames))]
+        public Task ReadDataStreamOfExternalAssetsNodeAsync(string testCaseName) =>
+            VerifyDataStreamOfTarUncompressedInternalAsync("node-tar", testCaseName, copyData: false);
+
+        [Theory]
+        [MemberData(nameof(GetRsTarTestCaseNames))]
+        public Task ReadDataStreamOfExternalAssetsRsAsync(string testCaseName) =>
+            VerifyDataStreamOfTarUncompressedInternalAsync("tar-rs", testCaseName, copyData: false);
+
+        [Theory]
+        [MemberData(nameof(GetGoLangTarTestCaseNames))]
+        public Task ReadCopiedDataStreamOfExternalAssetsGoLangAsync(string testCaseName) =>
+            VerifyDataStreamOfTarUncompressedInternalAsync("golang_tar", testCaseName, copyData: true);
+
+        [Theory]
+        [MemberData(nameof(GetNodeTarTestCaseNames))]
+        public Task ReadCopiedDataStreamOfExternalAssetsNodeAsync(string testCaseName) =>
+            VerifyDataStreamOfTarUncompressedInternalAsync("node-tar", testCaseName, copyData: true);
+
+        [Theory]
+        [MemberData(nameof(GetRsTarTestCaseNames))]
+        public Task ReadCopiedDataStreamOfExternalAssetsRsAsync(string testCaseName) =>
+            VerifyDataStreamOfTarUncompressedInternalAsync("tar-rs", testCaseName, copyData: true);
+
+        [Fact]
+        public async Task Throw_FifoContainsNonZeroDataSectionAsync()
+        {
+            await using MemoryStream archiveStream = GetTarMemoryStream(CompressionMethod.Uncompressed, "golang_tar", "hdr-only");
+            await using TarReader reader = new TarReader(archiveStream);
+            Assert.NotNull(await reader.GetNextEntryAsync());
+            Assert.NotNull(await reader.GetNextEntryAsync());
+            Assert.NotNull(await reader.GetNextEntryAsync());
+            Assert.NotNull(await reader.GetNextEntryAsync());
+            Assert.NotNull(await reader.GetNextEntryAsync());
+            Assert.NotNull(await reader.GetNextEntryAsync());
+            Assert.NotNull(await reader.GetNextEntryAsync());
+            Assert.NotNull(await reader.GetNextEntryAsync());
```

Comment:
nit: personally, I would use for loop here and add a comment explaining why we expect 8 entries

```suggestion
        for (int i = 0; i < 8; i++)
        {
            Assert.NotNull(await reader.GetNextEntryAsync());
        }
```